### ISSUE TITLE
ask about TEE shared memeory size.

### ISSUE THREAD

**Issue by SangsubLim** [OP-TEE/optee_os#6004]
Hello,
I ask about TEE shared memeory size.

In the GP Internal TEE documentation, it says:

> TEE_Client_API_Specification-V1.0.pdf
> 4.4.1. Configuration Settings
> 
> The value of TEEC_CONFIG_SHAREDMEM_MAX_SIZE is >= 0x80000, and it is written as follows in commnet.
> The maximum size of a single Shared Memory block, in bytes, of both API allocated and API registered memory.
> This version of the standard requires that this maximum size is greater than or equal to 512kB.
> In systems where there is no limit imposed by the Implementation then this definition should be defined to be the size of the address space.

It is understood that it should be possible to allocate more than 512KB through one shared memroy (TEEC_SharedMemory). Am I right?

The shared memory allocated by CA will be delivered to the TA side through the Invoke command. During the Invoke command, params can deliver up to 4 shared memories through memref.
If so, is it possible to share memory for at least 2MB (512kb*4) during the Invoke command if it conforms to the GP TEE standard?
In other words, I wonder if Targets that support OPTEE can share at least 2MB of memory with one invoke command.

As another question, I am conducting some experiments on the shared memory size in QEMU v8.

As far as I've tested, there is no problem with the invoke command up to 4MB (shared mem 1MB*4).
And you can run it with shared mem 4MB * 1.

If it is set to 8MB (shared mem 2MB*4), the following error occurs.

> TEEC_InvokeCommand failed with code 0xffff000c origin 0x4

Where can I check the size limit of shared memory?

BR,

**Comment by jenswi-linaro** [comment]
You could be running out of translation tables, one easy way of increasing them is to use a larger value for `CFG_NUM_THREADS`. See `PGT_CACHE_SIZE` in `core/arch/arm/include/mm/pgt_cache.h`.

**Comment by SangsubLim** [comment]
@jenswi-linaro 

Thank you for your reply.
Did I understand the following part of the question correctly?

In the GP Internal TEE documentation, it says:

TEE_Client_API_Specification-V1.0.pdf
4.4.1. Configuration Settings

> The value of TEEC_CONFIG_SHAREDMEM_MAX_SIZE is >= 0x80000, and it is written as follows in commnet.
> The maximum size of a single Shared Memory block, in bytes, of both API allocated and API registered memory.
> This version of the standard requires that this maximum size is greater than or equal to 512kB.
> In systems where there is no limit imposed by the Implementation then this definition should be defined to be the size of the address space.

It is understood that it should be possible to allocate more than 512KB through one shared memroy (TEEC_SharedMemory). Am I right?

The shared memory allocated by CA will be delivered to the TA side through the Invoke command. During the Invoke command, params can deliver up to 4 shared memories through memref.
If so, is it possible to share memory for at least 2MB (512kb*4) during the Invoke command if it conforms to the GP TEE standard?
In other words, I wonder if Targets that support OPTEE can share at least 2MB of memory with one invoke command.

BR,

**Comment by jenswi-linaro** [comment]
You're in my opinion trying to read out things that aren't specified.
The specification says a minimal max size of 512kB, but even that is in practice a best-effort size. You may need to tune OP-TEE depending on the size of used TAs and so on. OP-TEE doesn't impose explicit limits, but implicit limits as a consequence of various bottlenecks.
